### ISSUE TITLE
Using subword information in OOV token from fasttext in word embedding layer (keras/tensorflow)

### ISSUE THREAD

**Issue by datistiquo** [tensorflow/tensorflow#28524]
Serious, there is nothing on the internet I found in this way to handle OOV tokens with custom word vectors! There are some post, but all are saying using eg. fasttext (https://stackoverflow.com/questions/48395570/how-to-initialize-word-embeddings-for-out-of-vocabulary-word?rq=1 ) wih no explanation how to do it.
It should be somehow possibleto  bypass/hack around to assign a vector for OOVs?
But I struggle to find the proper position of code like in tf.nn.embedding_lookup where I can do it in a way like `if not in vocab get.fasttext.vector`

I have my own Fasttext model and trained with it a tensorflow classification model with a word embedding layer.

```
with tf.variable_scope('embeddings'):
            word_embeddings = tf.constant(self.embedding_mat, dtype=tf.float32, name="embedding")
            self.embedded_x1 = tf.nn.embedding_lookup(word_embeddings, self.x1)
            self.embedded_x2 = tf.nn.embedding_lookup(word_embeddings, self.x2)

```
But, I wonder how I can make use of the subword information of my model for OOV words? Since the word embedding layer operated via indices to look up word vectors and OOV words have no index. Even if a OOV token has a index how would I assign it the proper word vector to this OOV on the fly for an already trained model?

One understanding question: If I have a UNK token in my matrix but I did not used this UNK in training, can I assign a custom vector (eg. from fasttext) during prediction? So, for this UNK word which might be similar to a trained word such that the text is classified as same class? Then you could have also more UNK token like UNK1 and UNK2. 

I found two interesting posts, but I do't know if I understand that solution and if that is the solution...
https://stackoverflow.com/questions/45113130/how-to-add-new-embeddings-for-unknown-words-in-tensorflow-training-pre-set-fo

https://stackoverflow.com/questions/47022591/use-tensorflow-and-pre-trained-fasttext-to-get-embeddings-of-unseen-words?rq=1

Thanks in advance!

**Comment by achandraa** [comment]
This question is better asked on [StackOverflow](http://stackoverflow.com/questions/tagged//tensorflow) since it is not a bug or feature request. There is also a larger community that reads questions there and can provide support better for these type of issues. Thanks!

**Comment by datistiquo** [comment]
@achandraa I asked this too on stackoverflow but no answer. I do also get rarely an answer there...

**Comment by ymodak** [comment]
Closing this issue since its not Bug/Feature Request. [Stack Overflow](http://stackoverflow.com/questions/tagged/tensorflow) can be good platform to address this question. Thanks!

**Comment by mustfkeskin** [comment]
I have same problem with you
I replace embedding layer with DataGenerator
It's a bit slow but it works fine at least

`data_generator = DataGenerator(clean_text_list, batch_size=batch_size, shuffle=False)
`
```
def fasttext2vec(text):
    matrix = np.zeros((max_seq_length, word2vec.get_dimension()))
    for i, word in enumerate(text.split()):
        if i < max_seq_length:
            matrix[i] = word2vec.get_word_vector(word)     
    return matrix
```

`import numpy as np
import keras

class DataGenerator(keras.utils.Sequence):
    'Generates data for Keras'
    def __init__(self,
                 texts,
                 batch_size=256,
                 n_channels=1,
                 dim=(max_seq_length, word2vec.get_dimension()),
                 shuffle=True):
        'Initialization'
        self.dim = dim
        self.batch_size = batch_size
        self.n_channels = n_channels
        self.texts = texts
        self.shuffle = shuffle
        self.on_epoch_end()

    def __len__(self):
        'Denotes the number of batches per epoch'
        return int(np.floor(len(self.texts) / self.batch_size))

    def __getitem__(self, index):
        'Generate one batch of data'
        # Generate indexes of the batch
        indexes = self.indexes[index*self.batch_size:(index+1)*self.batch_size]

        # Find list of IDs
        texts_list = [self.texts[k] for k in indexes]

        # Generate data
        X = self.__data_generation(texts_list)

        return X

    def on_epoch_end(self):
        'Updates indexes after each epoch'
        self.indexes = np.arange(len(self.texts))
        if self.shuffle == True:
            np.random.shuffle(self.indexes)

    def __data_generation(self, texts):
        'Generates data containing batch_size samples' # X : (n_samples, *dim, n_channels)
        # Initialization
        text_matrix   = np.empty((self.batch_size, *self.dim))
        
        # Generate data
        for i, text in enumerate(texts):
            
            # Store sample
            text_matrix[i,] = vectorized_fastext2vec(text)

        return text_matrix`
